### PR TITLE
Add phone calls anomaly detection tutorial

### PR DESCRIPTION
**Pull request by guinao**
## Summary

Add the phone calls anomaly detection tutorial. This tutorial demonstrates how to use ML.net to do analysis on time series with seasonality. I.e, detect period and detect anomalies on the series. 

Preview Links

File Type | File Name | Published Url
-- | -- | --
Content | docs/machine-learning/tutorials/phone-calls-anomaly-detection.md | https://review.docs.microsoft.com/en-us/dotnet/machine-learning/tutorials/phone-calls-anomaly-detection?branch=pr-en-us-21845
Content | docs/machine-learning/tutorials/sales-anomaly-detection.md | https://review.docs.microsoft.com/en-us/dotnet/machine-learning/tutorials/sales-anomaly-detection?branch=pr-en-us-21845

### CODE REVIEW

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+Learn how to build an anomaly detection application for time series data. This tutorial creates a .NET Core console application using C# in Visual Studio 2019.
+
+In this tutorial, you learn how to:
+> [!div class="checklist"]
+>
+> * Load the data
+> * Detect period for a time series
+> * Detect anomaly for a periodical time series
+
+You can find the source code for this tutorial at the [dotnet/samples](https://github.com/dotnet/samples/tree/master/machine-learning/tutorials/ProductSalesAnomalyDetection) repository.
```
Comment (Youssef1313) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:20]: Wrong link?

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+## Prerequisites
+
+* [Visual Studio 2019 version 16.7.8 or later](https://visualstudio.microsoft.com/downloads/?utm_medium=microsoft&utm_source=docs.microsoft.com&utm_campaign=inline+link&utm_content=download+vs2019) with the ".NET Core cross-platform development" workload installed.
```
Comment (Youssef1313) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:24]: I think new articles are preferred to be compatible with .NET 5.0

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md (hunk 1 above)
Comment (luisquintanilla) [reply docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:20]: Once the PR in the samples repo is merged, then this link will work.

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+   using Microsoft.ML.Data;
+   ```
+
+4. Remove the existing class definition and add the following code, which has two classes `PhoneCallsData` and `PhoneCallsPrediction`, to the *PhoneCallsData.cs* file:
+
+    [!code-csharp[DeclareTypes](./snippets/phone-calls-anomaly-detection/csharp/PhoneCallsData.cs#DeclareTypes "Declare data record types")]
+
+    `PhoneCallsData` specifies an input data class. The [LoadColumn](xref:Microsoft.ML.Data.LoadColumnAttribute.%23ctor%28System.Int32%29) attribute specifies which columns (by column index) in the dataset should be loaded.
+
+    `PhoneCallsPrediction` specifies the prediction data class. For SR-CNN detector, the prediction is variable accroding to the [detect mode](https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.timeseries.srcnndetectmode?view=ml-dotnet) specified. In this sample we select the AnomalyAndMargin mode. It outputs seven columns. In most cases, `IsAnomaly`, `ExpectedValue`, `UpperBoundary` and `LowerBoundary` are informative enough. They tell you if a point is anomaly, the expected value of the point and the lower / upper boundary region of the point.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:88]:
```suggestion
    `PhoneCallsPrediction` specifies the prediction data class. For SR-CNN detector, the prediction depends on the [detect mode](xref:Microsoft.ML.TimeSeries.SrCnnDetectMode) specified. In this sample we select the `AnomalyAndMargin` mode. The output contains seven columns. In most cases, `IsAnomaly`, `ExpectedValue`, `UpperBoundary` and `LowerBoundary` are informative enough. They tell you if a point is an anomaly, the expected value of the point and the lower / upper boundary region of the point.
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+5. Add the following code to the line right above the `Main` method to specify those paths:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:90]:
```suggestion
5. Add the following code to the line right above the `Main` method to specify the path to your data file:
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+## Time series anomaly detection
+
+Time series anomaly detection is the process of detecting time-series data outliers; points on a given input time-series where the behavior isn't what was expected, or "weird". These anomalies can be typical indicative of some events of interest in the problem domain: a cyber-attack on user accounts, power outage, bursting RPS on a server, memory leak, etc.
+
+To find anomaly on time series, one should first determine the period of the series. Then, the time series can be decomposed into several components as `Y = T + S + R`, where `Y` is the original series, `T` is the trend component, `S` is the seasonal componnent and `R` is the residual component of the series. This step is called [decomposition](https://en.wikipedia.org/wiki/Decomposition_of_time_series). Finally, detection is performed on the residual component to find the anomalies. In ML.NET, The SR-CNN algorithm is an advanced and novel algorithm that is based on Spectral Residual (SR) and Convolutional Neural Network(CNN) to detect anomaly on time-series( Refer to [this](https://arxiv.org/pdf/1906.03821.pdf) paper for more details on this algorithm).
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:116]:
```suggestion
To find anomaly on time series, you should first determine the period of the series. Then, the time series can be decomposed into several components as `Y = T + S + R`, where `Y` is the original series, `T` is the trend component, `S` is the seasonal componnent and `R` is the residual component of the series. This step is called [decomposition](https://en.wikipedia.org/wiki/Decomposition_of_time_series). Finally, detection is performed on the residual component to find the anomalies. In ML.NET, The SR-CNN algorithm is an advanced and novel algorithm that is based on Spectral Residual (SR) and Convolutional Neural Network(CNN) to detect anomaly on time-series( Refer to the [Time-Series Anomaly Detection Service at Microsoft](https://arxiv.org/pdf/1906.03821.pdf) paper for more details on this algorithm).
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+1. Add the following code as the next line of the `Main()` method:
+
+    [!code-csharp[LoadData](./snippets/phone-calls-anomaly-detection/csharp/Program.cs#LoadData "loading dataset")]
+
+    The [LoadFromTextFile()](xref:Microsoft.ML.TextLoaderSaverCatalog.LoadFromTextFile%60%601%28Microsoft.ML.DataOperationsCatalog,System.String,System.Char,System.Boolean,System.Boolean,System.Boolean,System.Boolean%29) defines the data schema and reads in the file. It takes in the data path variables and returns an `IDataView`.
+
+## Time series anomaly detection
+
+Time series anomaly detection is the process of detecting time-series data outliers; points on a given input time-series where the behavior isn't what was expected, or "weird". These anomalies can be typical indicative of some events of interest in the problem domain: a cyber-attack on user accounts, power outage, bursting RPS on a server, memory leak, etc.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:114]:
```suggestion
Time series anomaly detection is the process of detecting time-series data outliers; points on a given input time-series where the behavior isn't what was expected, or "weird". These anomalies are typically indicative of some events of interest in the problem domain: a cyber-attack on user accounts, power outage, bursting RPS on a server, memory leak, etc.
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+
+To find anomaly on time series, one should first determine the period of the series. Then, the time series can be decomposed into several components as `Y = T + S + R`, where `Y` is the original series, `T` is the trend component, `S` is the seasonal componnent and `R` is the residual component of the series. This step is called [decomposition](https://en.wikipedia.org/wiki/Decomposition_of_time_series). Finally, detection is performed on the residual component to find the anomalies. In ML.NET, The SR-CNN algorithm is an advanced and novel algorithm that is based on Spectral Residual (SR) and Convolutional Neural Network(CNN) to detect anomaly on time-series( Refer to [this](https://arxiv.org/pdf/1906.03821.pdf) paper for more details on this algorithm).
+
+In this tutorial, you will see that these procedures can be completed in ML.net with simply two function invokes.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:118]:
```suggestion
In this tutorial, you will see that these procedures can be completed using two functions.
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+## Detect Period
+
+In the first step, we invoke the `DetectSeasonality` function to determine the period of the series.
+
+### Create the DetectPeriod() method
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:124]:
```suggestion
### Create the DetectPeriod method
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+
+The `DetectPeriod()` method:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:126]: I don't think this is needed

```suggestion
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+
+1. Create the `DetectPeriod()` method, just after the `Main()` method, using the following code:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:128]:
```suggestion
1. Create the `DetectPeriod` method, just below the `Main` method, using the following code:
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+
+    ```csharp
+    static void DetectPeriod(MLContext mlContext, IDataView phoneCalls)
+    {
+
+    }
+    ```
+
+2. Use the [DetectSeasonality](xref:Microsoft.ML.TimeSeriesCatalog.DetectSeasonality) function to detect period. Add it to the `DetectPeriod()` method with the following code:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:137]:
```suggestion
2. Use the [DetectSeasonality](xref:Microsoft.ML.TimeSeriesCatalog.DetectSeasonality) function to detect period. Add it to the `DetectPeriod` method with the following code:
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+
+    [!code-csharp[DetectSeasonality](./snippets/phone-calls-anomaly-detection/csharp/Program.cs#DetectSeasonality)]
+
+3. Display the period value by adding the following as the next line of code in the `DetectPeriod()` method:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:141]:
```suggestion
3. Display the period value by adding the following as the next line of code in the `DetectPeriod` method:
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+
+    [!code-csharp[DisplayPeriod](./snippets/phone-calls-anomaly-detection/csharp/Program.cs#DisplayPeriod)]
+
+4. Add the following call to the `DetectPeriod()`method in the `Main()` method:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:145]:
```suggestion
4. Add the following call to the `DetectPeriod` method in the `Main` method:
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+
+    [!code-csharp[CallDetectPeriod](./snippets/phone-calls-anomaly-detection/csharp/Program.cs#CallDetectPeriod)]
+
+
+### Period detection results
+
+Your results should be similar to the following.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:152]: Is the user supposed to run the application at this point? If so, I would specify it here.

```suggestion
Run the application. Your results should be similar to the following.
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+
+```console
+Period of the series is: 7.
+```
+
+## Detect Anomaly
+
+In this step, we use the [`SrCnnEntireDetector`](xref:Microsoft.ML.Transforms.TimeSeries.SrCnnEntireAnomalyDetector) to find anomalies.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:160]:
```suggestion
In this step, you use the [`SrCnnEntireDetector`](xref:Microsoft.ML.Transforms.TimeSeries.SrCnnEntireAnomalyDetector) to find anomalies.
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+
+### Create the DetectAnomaly() method
+
+1. Create the `DetectAnomaly()` method, just after the `Main()` method, using the following code:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:164]:
```suggestion
1. Create the `DetectAnomaly` method, just below the `DetectPeriod` method, using the following code:
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+
+```console
+Period of the series is: 7.
+```
+
+## Detect Anomaly
+
+In this step, we use the [`SrCnnEntireDetector`](xref:Microsoft.ML.Transforms.TimeSeries.SrCnnEntireAnomalyDetector) to find anomalies.
+
+### Create the DetectAnomaly() method
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:162]:
```suggestion
### Create the DetectAnomaly method
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+1. Create the `DetectAnomaly()` method, just after the `Main()` method, using the following code:
+
+    ```csharp
+    static void DetectAnomaly(MLContext mlContext, IDataView phoneCalls, int period)
+    {
+
+    }
+    ```
+
+2. Setup [SrCnnEntireAnomalyDetectorOptions](xref:Microsoft.ML.Transforms.TimeSeries.SrCnnEntireAnomalyDetectorOptions) in the `DetectAnomaly()` method with the following code:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:173]:
```suggestion
2. Setup [SrCnnEntireAnomalyDetectorOptions](xref:Microsoft.ML.Transforms.TimeSeries.SrCnnEntireAnomalyDetectorOptions) in the `DetectAnomaly` method with the following code:
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+
+    [!code-csharp[SetupSrCnnParameters](./snippets/phone-calls-anomaly-detection/csharp/Program.cs#SetupSrCnnParameters)]
+
+3. Detect anomaly by SR-CNN algorithm by adding the following line of code in the `DetectAnomaly()` method:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:177]:
```suggestion
3. Detect anomaly by SR-CNN algorithm by adding the following line of code in the `DetectAnomaly` method:
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+
+    [!code-csharp[DetectAnomaly](./snippets/phone-calls-anomaly-detection/csharp/Program.cs#DetectAnomaly)]
+
+4. Convert the output data view into a strongly typed `IEnumerable` for easier display using the `CreateEnumerable()`method with the following code:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:181]:
```suggestion
4. Convert the output data view into a strongly typed `IEnumerable` for easier display using the [`CreateEnumerable`](xref:Microsoft.ML.DataOperationsCatalog.CreateEnumerable%2A) method with the following code:
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 21]
```diff
@@ -0,0 +1,261 @@
...
+
+    [!code-csharp[CreateEnumerableForResult](./snippets/phone-calls-anomaly-detection/csharp/Program.cs#CreateEnumerableForResult)]
+
+5. Create a display header with the following code as the next line in the `DetectAnomaly()` method:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:185]:
```suggestion
5. Create a display header with the following code as the next line in the `DetectAnomaly` method:
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+    * `Anomaly` is the indicator of wheather each point is detected as anomaly.
+    * `ExpectedValue` is the estimated value of each point.
+    * `LowerBoundary` is the lowest value each point can be to be not anomaly.
+    * `UpperBoundary` is the highest value each point can be to be not anomaly.
+
+6. Iterate through the `predictions` `IEnumerable` and display the results with the following code:
+
+    [!code-csharp[DisplayAnomalyDetectionResults](./snippets/phone-calls-anomaly-detection/csharp/Program.cs#DisplayAnomalyDetectionResults)]
+
+7. Add the following call to the `DetectAnomaly()`method in the `Main()` method:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:201]:
```suggestion
7. Add the following call to the `DetectAnomaly` method in the `Main` method:
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 23]
```diff
@@ -0,0 +1,261 @@
...
+
+    [!code-csharp[CallDetectAnomaly](./snippets/phone-calls-anomaly-detection/csharp/Program.cs#CallDetectAnomaly)]
+
+## Anomaly detection results
+
+Your results should be similar to the following. During processing, messages are displayed. You may see warnings, or processing messages. Some messages have been removed from the following results for clarity.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:207]:
```suggestion
Run the application. Your results should be similar to the following. During processing, messages are displayed. You may see warnings, or processing messages. Some messages have been removed from the following results for clarity.
```

File: docs/machine-learning/tutorials/sales-anomaly-detection.md [hunk 24]
```diff
@@ -379,6 +379,6 @@ In this tutorial, you learned how to:
 
 ## Next steps
 
-Check out the Machine Learning samples GitHub repository to explore a Power Consumption Anomaly Detection sample.
+Check out the Machine Learning samples GitHub repository to explore a seasonality dat a anomaly detection sample.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/sales-anomaly-detection.md:382]:
```suggestion
Check out the Machine Learning samples GitHub repository to explore a seasonality data anomaly detection sample.
```

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 25]
```diff
@@ -0,0 +1,261 @@
...
+
+   ```csharp
+   using Microsoft.ML.Data;
+   ```
+
+4. Remove the existing class definition and add the following code, which has two classes `PhoneCallsData` and `PhoneCallsPrediction`, to the *PhoneCallsData.cs* file:
+
+    [!code-csharp[DeclareTypes](./snippets/phone-calls-anomaly-detection/csharp/PhoneCallsData.cs#DeclareTypes "Declare data record types")]
+
+    `PhoneCallsData` specifies an input data class. The [LoadColumn](xref:Microsoft.ML.Data.LoadColumnAttribute.%23ctor%28System.Int32%29) attribute specifies which columns (by column index) in the dataset should be loaded.
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:86]: For this section, would it be possible to expand a little bit more about what the the input columns represent. The timestamp is kind of self explanatory but there's not a lot of information on the value column

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md [hunk 26]
```diff
@@ -0,0 +1,261 @@
...
+
+    [!code-csharp[CreateMLContext](./snippets/phone-calls-anomaly-detection/csharp/Program.cs#CreateMLContext "Create the ML Context")]
+
+    The [MLContext class](xref:Microsoft.ML.MLContext) is a starting point for all ML.NET operations, and initializing `mlContext` creates a new ML.NET environment that can be shared across the model creation workflow objects. It's similar, conceptually, to `DBContext` in Entity Framework.
+
+### Load the data
+
+Data in ML.NET is represented as an [IDataView class](xref:Microsoft.ML.IDataView). `IDataView` is a flexible, efficient way of describing tabular data (numeric and text). Data can be loaded from a text file or from other sources (for example, SQL database or log files) to an `IDataView` object.
+
+1. Add the following code as the next line of the `Main()` method:
```
Comment (luisquintanilla) [inline docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:106]:
```suggestion
1. Add the following code as the next line of the `Main` method:
```

**Review by luisquintanilla — CHANGES_REQUESTED** [review]
Hi @guinao 

Thanks for this tutorial. It looks great! I added some suggestions. Most of them you should be able to commit directly.

File: docs/machine-learning/tutorials/phone-calls-anomaly-detection.md (hunk 2 above)
Comment (guinao) [reply docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:24]: Is there anything needs to change?

Comment (luisquintanilla) [reply docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:24]: @guinao can you just add a bullet point pointing out that they need to .NET 5 SDK. I believe it comes preinstalled with the latest version of VS, but just to make it clear that this is a .NET 5 sample.

Comment (luisquintanilla) [reply docs/machine-learning/tutorials/phone-calls-anomaly-detection.md:24]: No change, can you just make sure the version of VS2019 already comes with .NET 5 installed.

**Review by luisquintanilla — APPROVED** [review]
Thanks for your contribution @guinao! :shipit:

**Pull request merged**